File: java/src/main/java/com/agentharness/orchestration/Pipeline.java
package com.agentharness.orchestration;

import com.agentharness.Agent;
import com.agentharness.Harness;
import com.agentharness.model.AgentInput;
import com.agentharness.model.AgentOutput;
import com.agentharness.model.Decision;
import com.agentharness.model.DecisionAction;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Sequential Pipeline orchestration (harness-protocol.md §6.1).
 *
 * <p>Agents run in order. Each stage receives the prior stage's decision in its {@code context} under
 * the {@code "pipeline"} key, so a later stage can react to what came before. Every stage is invoked
 * through {@link Harness#invoke}, so each passes the confidence gate and tool registry individually
 * (O-1).
 *
 * <p>The pipeline short-circuits on the first {@code BLOCK} or {@code DEFER} (spec §6.1): once a stage
 * refuses or defers, later stages do not run — the safe action stands and no further side effects are
 * risked.
 */
public final class Pipeline {

    /** Actions that stop the pipeline (spec §6.1). */
    private static final Set<DecisionAction> SHORT_CIRCUIT =
            Set.of(DecisionAction.BLOCK, DecisionAction.DEFER);

    private final Harness harness;
    private final List<Agent> stages;

    public Pipeline(Harness harness, List<Agent> stages) {
        if (stages == null || stages.isEmpty()) {
            throw new IllegalArgumentException("a pipeline requires at least one stage");
        }
        this.harness = harness;
        this.stages = List.copyOf(stages);
    }

    public PipelineResult run(AgentInput request) {
        Map<String, AgentOutput> outputs = new LinkedHashMap<>();
        String shortCircuitedAt = null;
        AgentInput current = request;
        DecisionAction finalAction = null;

        for (Agent stage : stages) {
            AgentOutput output = harness.invoke(stage, current);
            outputs.put(stage.name(), output);
            finalAction = output.decision().action();
            if (SHORT_CIRCUIT.contains(finalAction)) {
                shortCircuitedAt = stage.name();
                break;
            }
            current = withPrior(request, stage.name(), output);
        }

        return new PipelineResult(finalAction, Map.copyOf(outputs), shortCircuitedAt);
    }

    /**
     * Return a new AgentInput carrying the prior stage's decision in {@code context.get("pipeline")}.
     * Scope and metadata are preserved verbatim (spec §2.1).
     */
    private static AgentInput withPrior(AgentInput base, String stageName, AgentOutput output) {
        Decision decision = output.decision();
        Map<String, Object> pipeline = new HashMap<>();
        pipeline.put("prior_stage", stageName);
        pipeline.put("prior_action", decision.action().name());
        pipeline.put("prior_confidence", decision.confidence());
        pipeline.put("prior_rationale", decision.rationale());

        Map<String, Object> context = new HashMap<>(base.context());
        context.put("pipeline", pipeline);
        return new AgentInput(base.tenantId(), base.userId(), context, base.metadata());
    }
}

File: bindings/python/src/halo_agent_harness/orchestration/pipeline.py
"""Sequential Pipeline orchestration (spec §6.1).

Agents run in order. Each stage receives the prior stage's decision in its ``context`` under the
``pipeline`` key, so a later stage can react to what came before. Every stage is invoked through
``Harness.invoke``, so each passes the confidence gate and tool registry individually (O-1).

The pipeline short-circuits on the first ``BLOCK`` or ``DEFER`` (spec §6.1): once a stage refuses or
defers, later stages do not run — the safe action stands and no further side effects are risked.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from ..core.agent import Agent
from ..core.harness import Harness
from ..core.model import AgentInput, AgentOutput, DecisionAction, reconcile

# Actions that stop the pipeline (spec §6.1) — the safer outcomes that must not be overridden by a
# later, more permissive stage.
_SHORT_CIRCUIT = frozenset({DecisionAction.BLOCK, DecisionAction.DEFER})


@dataclass(frozen=True)
class PipelineResult:
    """The outcome of a sequential pipeline run."""

    final_action: DecisionAction
    stage_outputs: dict[str, AgentOutput]
    short_circuited_at: str | None = None

    @property
    def reconciled_action(self) -> DecisionAction:
        """The safest action across every stage that actually ran (spec §3.3)."""
        return reconcile([o.decision.action for o in self.stage_outputs.values()])

    @property
    def any_auto_enforced(self) -> bool:
        return any(o.decision.auto_enforced for o in self.stage_outputs.values())


class Pipeline:
    """Runs agents sequentially, feeding each stage the prior stage's decision."""

    def __init__(self, harness: Harness, stages: list[Agent]) -> None:
        if not stages:
            raise ValueError("a pipeline requires at least one stage")
        self._harness = harness
        self._stages = list(stages)

    def run(self, request: AgentInput) -> PipelineResult:
        outputs: dict[str, AgentOutput] = {}
        short_circuited_at: str | None = None
        current = request

        for stage in self._stages:
            output = self._harness.invoke(stage, current)
            outputs[stage.name] = output
            action = output.decision.action
            if action in _SHORT_CIRCUIT:
                short_circuited_at = stage.name
                break
            current = _with_prior(request, stage.name, output)

        final_action = outputs[list(outputs)[-1]].decision.action
        return PipelineResult(
            final_action=final_action,
            stage_outputs=outputs,
            short_circuited_at=short_circuited_at,
        )


def _with_prior(base: AgentInput, stage_name: str, output: AgentOutput) -> AgentInput:
    """Return a new AgentInput carrying the prior stage's decision in ``context['pipeline']``.

    ``AgentInput`` is frozen; scope and metadata are preserved verbatim (spec §2.1).
    """
    decision = output.decision
    context = dict(base.context)
    context["pipeline"] = {
        "prior_stage": stage_name,
        "prior_action": decision.action.value,
        "prior_confidence": decision.confidence,
        "prior_rationale": decision.rationale,
    }
    return AgentInput(
        tenant_id=base.tenant_id,
        user_id=base.user_id,
        context=context,
        metadata=base.metadata,
    )
